### PR TITLE
feat(Project): New configuration make project and releases relationship more flexible for dependencies

### PR DESCRIPTION
**Pull request by hoangnt2**
# **How to enable this feature**

To use this function, please:

1. Build the source code and deploy.

2. Add config **enable.flexible.project.release.relationship=true** (/etc/sw360/sw360.properties) to enable the feature.

The following changes will work when **enable.flexible.project.release.relationship=true** only.

3. Use the migration script (**056_migrate_project_dependency_network.py**) we provided to mograte the database.

 Before you run the script, please change two places in the script:

 (1) Line 30: ```DRY_RUN = True``` -> ```DRY_RUN = False```

 (2) Line 32: ```COUCHSERVER = 'http://localhost:5984/'``` -> ```COUCHSERVER = 'http://admin:password@localhost:5984/'```

 ```admin``` and ```password``` should be your username and password for CouchDB.

# **The important information**

1. This feature has two modes - "ON" and "OFF". If the user doesn't enable this feature, everything will be the same as before (at the source code level).

2. For the "ON" mode, We add three new APIs and modify an old API to help users modify the dependency network of a project. But other old APIs will work as before. Users may not notice the change if they only use the old APIs. So all clients or tools working with sw360 by APIs will work as before as well. For example, our new API 3.3.36 (Creating a project with dependencies network) will allow the user to create a new project with a dependency network, but the user can still use the old API 3.3.20 (Creating a project) to register a project with some releases (i.e. direct dependencies) linked to itself.

3. For the "OFF" mode, we add a new field "releaseRelationNetwork" in the database to store the dependency network. But we still keep the old filed "releaseIdToUsage" which stores the "Linked Release". What's more, we always sync these two fields. So when the user switches back to "ON" mode, no data will be lost.

# **1. Introduction**

This manual aims to introduce the Dependency Network feature.  Refer issue: #1538 
The dependency network feature is a new function to make the dependency management of a project more flexible by allowing the users to customize the dependency graphs of their projects.
To help the reviewers test and review this new feature more quickly and easily, **Toshiba** makes this manual.

# **2. How to test?**
This feature modify the GUI of the “Linked Releases And Projects” on the “project edits” page.
Now the “Linked Releases” table could show all dependencies of a project (both direct and transitive ones). Users can modify these dependencies as well.

edit

## **2.1. The changes of edit project GUI**
In this section, we will introduce the changes in GUI behaviors. We modified or added 5 sub-functions below: 
 
#### **a. Modify the “Add Releases” button: This button will add a direct dependency (release) in the dependency graph of this project.**

add-release

#### **b. A new icon button to add a dependency (release) to another dependency (release) in the dependency graph. Note that this dependency added is seen as the transitive dependency of this project.**

add-child

#### **c. A new icon button to load the default dependency graph of a dependency (release) by importing the dependency information stored on the component page. Note that this button will load all dependencies (both direct and transitive ones) of the corresponding dependency (release).**

default

#### **d. The combo box allows the user to modify the version of a dependency.**

select-version

#### **e. The “Check Dependency Network” button will compare and show the different dependency information which is not consistent with the default one stored on the component page by highlighting them. The inconsistency usually happens after users modified the dependency graph or imported an old project.**

check-not-match

## **2.2. The changes of view detail project GUI**
The license Clearing tab will work the same as before. But it will show the data from the dependency network we modified on the edit page.

view-page

### New Rest APIs

**a. 3.3.35. Get a single project with dependencies network**

The response will include the dependencyNetwork field(It will show the dependency network of project (direct and indirect releases)):
```
{
  "name" : "Emerald Web",
  "dependencyNetwork": [
      {
          "releaseId": "9efc5766cd0c41d4a40547b99f5b91ac",
          "releaseLink": [
              {
                  "releaseId": "3bed97a1c7ac4c32846ef4be985b648c",
                  "releaseLink": [
                      {
                          "releaseId": "6a8250852362462095c57535294039e4",
                          "releaseLink": [],
                          "releaseRelationship": "TO_BE_REPLACED",
                          "mainlineState": "PHASEOUT",
                          "comment": "Test Comment",
                          "createOn": "2023-05-15",
                          "createBy": "admin@sw360.org"
                      }
                  ],
                  "releaseRelationship": "INTERNAL_USE",
                  "mainlineState": "OPEN",
                  "comment": "Test Comment",
                  "createOn": "2023-05-15",
                  "createBy": "admin@sw360.org"
              }
          ],
          "releaseRelationship": "STATICALLY_LINKED",
          "mainlineState": "MAINLINE",
          "comment": "Test Comment",
          "createOn": "2023-05-15",
          "createBy": "admin@sw360.org"
      },
      {
          "releaseId": "f1d860e7576a44798ee3daff57a3a886",
          "releaseLink": [],
          "releaseRelationship": "OPTIONAL",
          "mainlineState": "OPEN",
          "comment": "Test Comment",
          "createOn": "2023-05-15",
          "createBy": "admin@sw360.org"
      }
  ]
}
```

**b. 3.3.36. Creating a project with dependencies network**

If the **dependencyNetwork** field is included in the request body, a dependency network will be registered for the project.

- Simple example request (modify releaseIds to the existing release ids in sw360):

```
{
    "name": "TestProject1",
    "dependencyNetwork": [
        {
            "releaseId": "9efc5766cd0c41d4a40547b99f5b91ac",
            "releaseLink": [
                {
                    "releaseId": "3bed97a1c7ac4c32846ef4be985b648c",
                    "releaseLink": [
                    ],
                    "releaseRelationship": "INTERNAL_USE",
                    "mainlineState": "OPEN",
                    "comment": "Test Comment",
                    "createOn": "2023-05-15",
                    "createBy": "admin@sw360.org"
                }
            ],
            "releaseRelationship": "STATICALLY_LINKED",
            "mainlineState": "MAINLINE",
            "comment": "Test Comment",
            "createOn": "2023-05-15",
            "createBy": "admin@sw360.org"
        },
        {
            "releaseId": "f1d860e7576a44798ee3daff57a3a886",
            "releaseLink": [],
            "releaseRelationship": "OPTIONAL",
            "mainlineState": "OPEN",
            "comment": "Test Comment",
            "createOn": "2023-05-15",
            "createBy": "admin@sw360.org"
        }
    ]
}
```

**c. 3.3.37. Update a project with dependencies network**

Same request body as "Creating a project with dependencies network".

### Modify Rest APIs
**a. 3.5.18. Delete a release**

Cannot delete release if it is existing in the dependency network of any project

## **2.4 The change in database and migrate data**
- Created new field in project to save the dependency network: **releaseRelationNetwork**
- Create **056_migrate_project_dependency_network.py** to migrate data from **releaseIdToUsage** to **releaseRelationNetwork**

## **2.5 Test this feature with component.visibility.restriction.enabled option (refer #1357)**
- Case 1: component.visibility.restriction.enabled=true
- Case 2: component.visibility.restriction.enabled=false

# **3. How to review?**
This chapter introduces which file is modified to achieve the corresponding function.

## **3.1. Rest API**

**Table 1: RestAPI functions**
| Rest API function  | File Path | Function name|
| ----------- | ----------- |----------- |
| 3.3.34. Get a single project with dependencies network      | rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java|getProjectWithNetwork()|
| 3.3.35. Creating a project with dependencies network     | rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java|createProjectWithNetwork()|
| 3.3.36. Update a project with dependencies network      | rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/project/ProjectController.java|patchProjectWithNetwork()|
| 3.5.18. Delete a release    | rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/Sw360ReleaseService.java|deleteRelease()|
| 3.3.20. Creating a project    | rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/Sw360ReleaseService.java|createProject()|
| 3.3.22. Update a project    | rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/Sw360ReleaseService.java|patchProject()|

## **3.2. GUI**
**Table 2: UI functions**
| GUI function  | File Path |
| ----------- | ----------- |
| Project edit page (Dependency network tab)     |  - frontend/sw360-portlet/src/main/resources/META-INF/resources/html/projects/edit.jsp <br>- frontend/sw360-portlet/src/main/resources/META-INF/resources/html/utils/includes/editLinkedReleasesInNetwork.jspf <br>- frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/LinkedReleasesAndProjectsAwarePortlet.java <br> - frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java|
| Project detail page (Dependency network tab)   |  - frontend/sw360-portlet/src/main/resources/META-INF/resources/html/projects/includes/detailOverview.jspf <br> - frontend/sw360-portlet/src/main/resources/META-INF/resources/html/projects/includes/projects/dependencyNetwork.jsp <br>- frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java|
|  -  Components detail page: <br>  +     Used By Project table     |   frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java|
|  -  Releases detail page: <br>  +     Used By Project table     |   frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/components/ComponentPortlet.java|

Reference documents:
[PR.for.Dependency.network.feature.docx](https://github.com/eclipse-sw360/sw360/files/11483658/PR.for.Dependency.network.feature.docx)

[A.New.Dependency.Management.Function.for.SW360_qiu.pdf](https://github.com/eclipse-sw360/sw360/files/11483696/A.New.Dependency.Management.Function.for.SW360_qiu.pdf)

### CODE REVIEW

**Comment by KoukiHama** [comment]
@shi9qiu @tienlee 
Please follow Anupam comments.

**Comment by tienlee** [comment]
Thanks for your comments. I will discuss with Qiu and give feedback you tomorrow.

**Comment by KoukiHama** [comment]
@shi9qiu any update?

**Comment by shi9qiu** [comment]
Hi @ag4ums ,

Thank you for your comments.

I have discussed this with Tien-san, please allow me to discuss more about this issue here.

We proposed this function with two modes - ON and OFF. Changes only happen when the user enables this function. If not, everything will be the same as before (at the source code level).

I would like to begin with the second concern - "not seeing an option to view only the direct dependencies separately". It's easy for us to develop a new feature to show the direct dependencies only based on the developed one. For example, hide the transitive dependencies when the user clicks a button. So this would not be a problem.

But we would like to discuss more on the first concern - "it will break all existing clients". Actually, besides the four APIs we modified and the dependency part in GUI, all other APIs and functions can be used as before. To achieve this, we add a new field (releaseRelationNetwork) to store the dependency network (direct dependencies and transitive dependencies) in the project and keep it in sync with the releaseIdToUsage field. The releaseIdToUsage field is not replaced by the new one and still be there. That's also the reason why other functions will not be affected. Besides, We also create a migration script to convert the data from the releaseIdToUsage field to the releaseRelationNetwork field. It will help users to switch between two modes.

I think maybe we use a bad name "Dependency Network" for this new tag. It looks like we drop the "License Clearing" tab and the function in this tab. But actually, the two tabs are almost the same. All functions are kept as before. The only difference is the new tab can show and edit the transitive dependency. So maybe we can still use the name "License Clearing",  and the user may even not notice the change.

I don't know if my reply answers your question and resolves your concerns. Maybe we misunderstand what you mean by "existing clients". But we are so glad to hear more from you and waiting to solve anything you are worrying about.

Thank you!

**Comment by ag4ums** [comment]
@shi9qiu 

> We proposed this function with two modes - ON and OFF. Changes only happen when the user enables this function. If not, everything will be the same as before (at the source code level).
 `ON/OFF mode is good, so user can choose.`

> I would like to begin with the second concern - "not seeing an option to view only the direct dependencies separately". It's easy for us to develop a new feature to show the direct dependencies only based on the developed one. For example, hide the transitive dependencies when the user clicks a button. So this would not be a problem.
`I think this is an important option to keep UI consistent.` 

> But we would like to discuss more on the first concern - "it will break all existing clients". Actually, besides the four APIs we modified and the dependency part in GUI, all other APIs and functions can be used as before. To achieve this, we add a new field (releaseRelationNetwork) to store the dependency network (direct dependencies and transitive dependencies) in the project and keep it in sync with the releaseIdToUsage field. The releaseIdToUsage field is not replaced by the new one and still be there. That's also the reason why other functions will not be affected. Besides, We also create a migration script to convert the data from the releaseIdToUsage field to the releaseRelationNetwork field. It will help users to switch between two modes.

`let me try to explain it, I tried to used ort analysis to create project and component, yet to use other rest client/s like capycli, and found that the projects and releases are getting created but releases are not linked to projects, as a result they don’t appear in the UI - Dependency Network tab. guessing other clients will also show the same behavior. 
tried to migrate the old project using the script but for migrated projects  'Dependency Network' tab is empty. (guessing migration issue)`

 > I think maybe we use a bad name "Dependency Network" for this new tag. It looks like we drop the "License Clearing" tab and the function in this tab. But actually, the two tabs are almost the same. All functions are kept as before. The only difference is the new tab can show and edit the transitive dependency. So maybe we can still use the name "License Clearing", and the user may even not notice the change.
`on the naming I think 'License clearing' is more meaning full here as this tab primarily helps with the 'license clearing', Depenendy Netwok' naming somehow feels incomplete as this is does not give you a complete network of the dependencies but only dependencies for a project.`  

So, coming back to the first point that enable this option works well, that’s good, so user can choose to enable it but I am afraid once user migrated to this new option to use dependency network functionality will they be able to switchback.... though not sure if the switching back is required, but what worries me.... if this feature is breaking the existing rest-client/s, and if someone do changes at the client side to support 'Dependency Network', that should not break the current flow(as I see the otherway its breaking now).
Let me know your opinion here... how we can resolve this.

**Comment by shi9qiu** [comment]
Hi @ag4ums ,

Thank you very much for your kind explanation. I understand your concern. 

Please give me some time, I would like to investigate this issue and discuss with @tienlee to consider how to resolve this. I will share our idea with you as soon as we get it. Thanks!

**Comment by ag4ums** [comment]
@shi9qiu, @tienlee, @hoangnt2 if you take care of the 3rd point about the API then it should be fine.

**Comment by shi9qiu** [comment]
Hi, @ag4ums 

We have finished a new version and pushed the source code to this PR. This new version is developed based on what we discussed above. Could you help to review it again?

The biggest change of this new version is that we don't change any existing APIs so they can be used as before for the "ON" mode. Instead, We added 3 new APIs related to the dependency network. We also updated the description of this PR to explain this.

If you have any questions or concerns, please feel free to contact us anytime!

**Comment by KoukiHama** [comment]
we plan to add some test cases for specific options in sw360 properties.

**Comment by hoangnt2** [comment]
@GMishx, thanks for your comments. I updated source code following your comments

File: frontend/sw360-portlet/bnd.bnd [hunk 1]
```diff
@@ -72,7 +72,9 @@ Import-Package: \
 	com.cloudant.*, \
 	org.osgi.service.cm, \
 	org.osgi.service.component.*, \
-	org.slf4j
+	org.slf4j, \
+    com.fasterxml.jackson.core.*,\
```
Comment (smrutis1) [inline frontend/sw360-portlet/bnd.bnd:76]: Indentation please

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/utils/ajax/linkedReleaseInNetwork.jsp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+<jsp:useBean id="releasesInNetwork" type="java.util.List<org.eclipse.sw360.datahandler.thrift.components.ReleaseLink>"  scope="request"/>
+<core_rt:set var="mainlineStateEnabledForUserRole" value='<%=PortalConstants.MAINLINE_STATE_ENABLED_FOR_USER%>'/>
+<core_rt:forEach items="${releasesInNetwork}" var="releaseLink" varStatus="loop">
+
+    <core_rt:choose>
+        <core_rt:when test="${releaseLink.accessible}">
+            <core_rt:set var="uuid" value="${releaseLink.id}"/>
+            <tr id="releaseLinkRow${uuid}" parent-node="${releaseLink.parentNodeId}" data-layer="${releaseLink.layer}" data-index="${releaseLink.index}">
+                <td class="form-group">
+                    <div style="font-size:1rem; display:flex; width:100%; border: none;">
```
Comment (smrutis1) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/utils/ajax/linkedReleaseInNetwork.jsp:33]: Kindly try to remove inline css

**Review by smrutis1 — CHANGES_REQUESTED** [review]
Kindly look into some minor comments

File: frontend/sw360-portlet/bnd.bnd (hunk 1 above)
Comment (hoangnt2) [reply frontend/sw360-portlet/bnd.bnd]: Fixed

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 3]
```diff
@@ -2969,4 +2969,45 @@ public ByteBuffer getComponentReportDataStream(User user, boolean extendedByRele
...
+        return Collections.singletonList(dependencyNetwork);
+    }
+
+    private ReleaseNode getReleaseNodes(ReleaseNode releaseNode, User user) {
+        Release releaseById = null;
+        try {
+            releaseById = getAccessibleRelease(releaseNode.getReleaseId(), user);
+            List<Release> releaseList = new ArrayList<>();
+            if (releaseById.getReleaseIdToRelationship() != null) {
+                releaseList = getAccessibleReleases(releaseById.getReleaseIdToRelationship().keySet().stream().collect(Collectors.toSet()), user);
```
[the remark this answers is not in the document]
Comment (hoangnt2) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: fixed

File: libraries/exporters/bnd.bnd [hunk 4]
```diff
@@ -31,7 +31,8 @@ Import-Package: \
 	org.eclipse.sw360.commonIO.*, \
 	org.eclipse.sw360.datahandler.*, \
 	org.xml.sax.*, \
-	org.w3c.dom
+	org.w3c.dom, \
+    com.fasterxml.jackson.*
```
[the remark this answers is not in the document]
Comment (hoangnt2) [reply libraries/exporters/bnd.bnd]: fixed

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 5]
```diff
@@ -2969,4 +2969,45 @@ public ByteBuffer getComponentReportDataStream(User user, boolean extendedByRele
...
+            releaseNode.setCreateOn(SW360Utils.getCreatedOn());
+            releaseNode.setCreateBy(user.getEmail());
+            releaseNode.setComment("");
+            for (Release release : releaseList) {
+                ReleaseNode node = new ReleaseNode(release.getId());
+                node.setMainlineState(MainlineState.OPEN.toString());
+                node.setReleaseRelationship(ReleaseRelationship.CONTAINED.toString());
+                node.setComment("");
+                node.setCreateOn(SW360Utils.getCreatedOn());
+                node.setCreateBy(user.getEmail());
```
[the remark this answers is not in the document]
Comment (hoangnt2) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java:3003]: Creator should be the current user

**Comment by hoangnt2** [comment]
@smrutis1, I updated source code following your comments, please help me to check them

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/utils/ajax/linkedReleaseInNetwork.jsp (hunk 2 above)
Comment (hoangnt2) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/utils/ajax/linkedReleaseInNetwork.jsp]: > Looks like the inline CSS still there.

@smrutis1, fixed

**Comment by KoukiHama** [comment]
Thank you @smrutis1  for review codes
Now we can merge this PR.

**Pull request merged**